Mathlib/AlgebraicGeometry/PullbackCarrier.lean

```diff
@@ -0,0 +1,356 @@
+/-
+Copyright (c) 2024 Qi Ge, Christian Merten, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Qi Ge, Christian Merten, Andrew Yang
+-/
+import Mathlib.Algebra.Category.Ring.LinearAlgebra
+import Mathlib.AlgebraicGeometry.ResidueField
+
+/-!
+# Underlying topological space of fibre product of schemes
+
+Let `f : X ⟶ S` and `g : Y ⟶ S` be morphisms of schemes. In this file we describe the underlying
+topological space of `pullback f g`, i.e. the fiber product `X ×[S] Y`.
+
+## Main results
+
+- `AlgebraicGeomerty.Scheme.Pullback.carrierEquiv`: The bijective correspondence between the points
+  of `X ×[S] Y` and pairs `(z, p)` of triples `z = (x, y, s)` with `f x = s = g y` and
+  prime ideals `q` of `κ(x) ⊗[κ(s)] κ(y)`.
+- `AlgebraicGeometry.Scheme.Pullback.exists_preimage`: For every triple `(x, y, s)` with
+  `f x = s = g y`, there exists `z : X ×[S] Y` lying above `x` and `y`.
+
+We also give the ranges of `pullback.fst`, `pullback.snd` and `pullback.map`.
+
+-/
+
+open CategoryTheory Limits TopologicalSpace LocalRing TensorProduct
+
+noncomputable section
+
+namespace AlgebraicGeometry.Scheme.Pullback
+
+universe u
+
+variable {X Y S : Scheme.{u}} (f : X ⟶ S) (g : Y ⟶ S)
+
+/-- A `Triplet` over `f : X ⟶ S` and `g : Y ⟶ S` is a triple of points `x : X`, `y : Y`,
+`s : S` such that `f x = s = f y`. -/
+structure Triplet where
```

Comment:
I'm slightly sad that this doesn't have a more descriptive name, but `Base` is all I can think of, which is no better.